console-subscriber/src/visitors.rs

```diff
@@ -0,0 +1,359 @@
+                                {
+                                    let op = match op.as_str() {
+                                        Self::UPDATE_OP_ADD => Some(AttributeUpdateOp::Add),
+                                        Self::UPDATE_OP_SUB => Some(AttributeUpdateOp::Sub),
+                                        Self::UPDATE_OP_OVR => Some(AttributeUpdateOp::Ovr),
+                                        _ => None,
+                                    };
+
+                                    return op.map(|op| AttributeUpdate {
+                                        name: attr_name,
+                                        val: AttributeUpdateValue::Numeric { val, op, unit },
+                                    });
+                                }
+                                None
+                            });
+
+                    let text_updates =
+                        self.state_text_attrs.into_iter().map(|(attr_name, attr)| {
+                            AttributeUpdate {
+                                name: attr_name,
+                                val: AttributeUpdateValue::Text(attr),
+                            }
+                        });
+
+                    let attrs = numeric_updates.chain(text_updates).collect();
+                    return Some(ResourceOpData::StateUpdate { op_name, attrs });
+                }
+            }
+        }
+        None
+    }
+}
+
+struct MatchPart {
+    attr_name: String,
+    part_type: MatchPartType,
+}
+
+enum MatchPartType {
+    OpType(String),
+    Unit(String),
+    Text(String),
+}
+
+impl Visit for ResourceOpVisitor {
+    fn record_debug(&mut self, _: &field::Field, _: &dyn std::fmt::Debug) {}
+
+    fn record_str(&mut self, field: &tracing_core::Field, value: &str) {
+        let extract_partial_attr = || -> Option<MatchPart> {
+            let mut parts = field.name().split('_');
+            let p0 = parts.next();
+            let p1 = parts.next();
+            let p2 = parts.next();
+            if let (Some(p0), Some(p1), Some(p2)) = (p0, p1, p2) {
+                if p0 == Self::OP_STATE_FIELD_PREFIX {
+                    if p2 == Self::OP_STATE_FIELD_TYPE_UNIT {
+                        return Some(MatchPart {
+                            attr_name: p1.into(),
+                            part_type: MatchPartType::Unit(value.into()),
+                        });
+                    } else if p2 == Self::OP_STATE_FIELD_TYPE_VALUE {
+                        return Some(MatchPart {
+                            attr_name: p1.into(),
+                            part_type: MatchPartType::Text(value.into()),
+                        });
+                    } else if p2 == Self::OP_STATE_FIELD_TYPE_OP {
+                        return Some(MatchPart {
+                            attr_name: p1.into(),
+                            part_type: MatchPartType::OpType(value.into()),
+                        });
+                    }
```

Comment:
nit/style: this looks like
```suggestion
                   match p2 {
                        Self::OP_STATE_FIELD_TYPE_UNIT => return Some(MatchPart {
                            attr_name: p1.into(),
                            part_type: MatchPartType::Unit(value.into()),
                        }),
                        Self::OP_STATE_FIELD_TYPE_VALUE => return Some(MatchPart {
                            attr_name: p1.into(),
                            part_type: MatchPartType::Text(value.into()),
                        }),
                        Self::OP_STATE_FIELD_TYPE_OP => return Some(MatchPart {
                            attr_name: p1.into(),
                            part_type: MatchPartType::OpType(value.into()),
                        }),
                        _ => {}
                    }
```